### PR TITLE
Use a vector instead of a set to represent clp-s schemas

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR replaces the in-memory representation of clp-s schemas with a class that wraps std::vector.

This new Schema class allows for schemas to contain a region of nodes ordered by mst node id, followed by a region of unordered nodes. This will be useful if we start supporting more generic parsers for some types of data which need to define their own order, or potentially if we change how we treat arrays.

This implementation still requires all of the node ids within a schema to be unique, though this restriction could be lifted if we make some changes to ParsedRecord and make some changes in Schema Matching.

Besides that, this change improves compression speed by about 10% for some datasets.

# Validation performed
- validated that compressed archives are identical for some datasets before and after this change
- benchmarked that compression speed is ~10% faster for some datasets
- benchmarked that search performance has no significant change

### CODE REVIEW

File: components/core/src/clp_s/ReaderUtils.hpp
Comment (LinZhihao-723) [inline components/core/src/clp_s/ReaderUtils.hpp:20]: If we add a `Schema` class in this PR, why are we still storing a vector here? I understand that when reading from the archive, we no longer need to care about the ordering, but this can be confusing since `Schema` is only used for compression and `std::vector<int32_t>` is for search/decompression. I have tried a little bit and here's my purposal to replace the vector by `Schema` where it's necessary:
1. add `[]` operator to the `Schema` class so that it can be indexed as a vector
2. when reading from the archive to a shema, instead of calling `push_back` we can call `insert_unordered` (since the order doesn't matter)
3. add `data()` method to get the underlying raw vector when passing the schema as a column

File: components/core/src/clp_s/ReaderUtils.cpp
Comment (LinZhihao-723) [inline components/core/src/clp_s/ReaderUtils.cpp:106]:
```suggestion
        auto& schema = schemas[schema_id];
```
So that we don't need to care about the underlying type.

File: components/core/src/clp_s/SchemaMap.cpp
Comment (LinZhihao-723) [inline components/core/src/clp_s/SchemaMap.cpp:7]: Even though this function is not a part of the PR, but I would suggest to rewrite it in the following way:
```
int32_t SchemaMap::add_schema(Schema const& schema) {
    auto const schema_it = m_schema_map.find(schema);
    if (m_schema_map.end() == schema_it) {
        return schema_it->second;
    }
    m_schema_map.emplace(schema, m_current_schema_id);
    return m_current_schema_id++;
}
```
Reasons:
1. const correctness on the iterator
2. Yoda condition
3. Trying to avoid inner return of an if-else statement (as clang-tidy would also alert)
4. Avoid using `[]` to insert element to a map/unordered_map

File: components/core/src/clp_s/SchemaMap.hpp
Comment (LinZhihao-723) [inline components/core/src/clp_s/SchemaMap.hpp:12]:
```suggestion
    using schema_map_t = std::map<Schema, int32_t>;
```

File: components/core/src/clp_s/SchemaMap.cpp
Comment (gibber9809) [reply components/core/src/clp_s/SchemaMap.cpp:7]: Makes sense to me, I'll change it.

Comment (LinZhihao-723) [reply components/core/src/clp_s/SchemaMap.cpp:7]: oh yeah, sorry for the mistake :D

File: components/core/src/clp_s/ReaderUtils.hpp
Comment (gibber9809) [reply components/core/src/clp_s/ReaderUtils.hpp]: I made changes to use Schema on the reader side as well, but I didn't end up having to add the `[]` or `data()` methods to get things working. The iterators on Schema satisfied the current uses without any additional code change.

Do you think we should still add these methods in case we want them later, or is this good for now?

**Review by LinZhihao-723 — DISMISSED** [review]

File: components/core/src/clp_s/Schema.hpp
Comment (kirkrodrigues) [inline components/core/src/clp_s/Schema.hpp:1]: We try to use trailing return types in new files to agree with our upcoming clang-tidy settings.

**Review by kirkrodrigues — COMMENTED** [review]
Won't hold up the PR for these comments, but would be good if you could fix them before we merge.

File: components/core/src/clp_s/Schema.hpp
Comment (gibber9809) [reply components/core/src/clp_s/Schema.hpp:56]: You would think, but surprisingly that doesn't seem to work with the for (auto item: schema) syntax.

File: components/core/src/clp_s/Schema.hpp
Comment (gibber9809) [reply components/core/src/clp_s/Schema.hpp:14]: I'm going to remove this schema_t and just have the "append a list of mst node ids" signature accept an entire Schema object.

File: components/core/src/clp_s/Schema.hpp
Comment (kirkrodrigues) [reply components/core/src/clp_s/Schema.hpp]: Oh I see, it's because the ranged-based for seems to look for a method named `begin` and doesn't consider `cbegin`. The way `std::vector` solves this is by defining a non-const `begin`, const `begin`, and const `cbegin`. Can we do the same here?

File: components/core/src/clp_s/Schema.hpp
Comment (kirkrodrigues) [inline components/core/src/clp_s/Schema.hpp:13]:
```suggestion
 * Internally, the schema is represented by a vector where the first m_num_ordered entries are
 * ordered by MST node ID, and the following entries are allowed to have arbitrary order.
 ```

File: components/core/src/clp_s/Schema.hpp
Comment (kirkrodrigues) [inline components/core/src/clp_s/Schema.hpp:16]:
```suggestion
 * In the current implementation of clp-s, MST node IDs in a schema must be unique, so the caller
 * is responsible for not inserting duplicate MST nodes into a schema. Future versions of clp-s will
```

File: components/core/src/clp_s/Schema.hpp
Comment (kirkrodrigues) [inline components/core/src/clp_s/Schema.hpp:32]:
```suggestion
     * Inserts another schema into the unordered region of the schema, maintaining that Schema's
```

File: components/core/src/clp_s/Schema.hpp
Comment (kirkrodrigues) [inline components/core/src/clp_s/Schema.hpp:56]:
```suggestion
     * @return iterator to the end of the underlying schema
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp_s/Schema.hpp
Comment (gibber9809) [reply components/core/src/clp_s/Schema.hpp]: Sure, I'll add cbegin/cend as well

**Review by kirkrodrigues — APPROVED** [review]

**Pull request merged**
